src/sensors.cpp

```diff
@@ -59,80 +59,115 @@ void Sensors::init()
   {
     rf_.state_manager_.set_error(StateManager::ERROR_UNCALIBRATED_IMU);
   }
+  next_sensor_to_update_ = 0;
 }
 
 
 bool Sensors::run(void)
 {
   // First, check for new IMU data
-  bool new_imu_data = update_imu();
-
-
-  // Now, Look for disabled sensors while disarmed (poll every 0.5 seconds)
-  // These sensors need power to respond, so they might not have been
-  // detected on startup, but will be detected whenever power is applied
-  // to the 5V rail.
-  if (!rf_.state_manager_.state().armed)
+  if (update_imu())
   {
-    uint32_t now = rf_.board_.clock_millis();
-    if (now > (last_time_look_for_disarmed_sensors + 500))
-    {
-      last_time_look_for_disarmed_sensors = now;
-//      if (!rf_.board_.sonar_present())
-//      {
-//        if (rf_.board_.sonar_check())
-//        {
-          //          mavlink_log_info("FOUND SONAR", NULL);
-//          volatile int debug = 1;
-//        }
-//      }
-      if (!rf_.board_.diff_pressure_present())
-      {
-        if (rf_.board_.diff_pressure_check())
-        {
-//                    mavlink_log_info("FOUND DIFF PRESS", NULL);
-        }
-      }
-    }
+    return true;
   }
-
-
-  // Update whatever sensors are available
-  if (rf_.board_.baro_present())
+  else
   {
-    rf_.board_.baro_read(&data_.baro_altitude, &data_.baro_pressure, &data_.baro_temperature);
+    if (!rf_.state_manager_.state().armed)
+      look_for_disabled_sensors();
+
+    // Update other sensors
+    update_other_sensors();
+    return false;
   }
+}
 
-  if (rf_.board_.diff_pressure_present())
+
+void Sensors::update_other_sensors()
+{
+  uint32_t now = rf_.board_.clock_millis();
+  switch (next_sensor_to_update_)
   {
+  case 0:
     if (rf_.board_.baro_present())
     {
-      rf_.board_.diff_pressure_set_atm(data_.baro_pressure);
+      rf_.board_.baro_read(&data_.baro_altitude, &data_.baro_pressure, &data_.baro_temperature);
     }
-    rf_.board_.diff_pressure_read(&data_.diff_pressure, &data_.diff_pressure_temp, &data_.diff_pressure_velocity);
+    break;
+  case 1:
+    if (rf_.board_.diff_pressure_present())
+    {
+      if (rf_.board_.baro_present())
+      {
+        rf_.board_.diff_pressure_set_atm(data_.baro_pressure);
+      }
+      rf_.board_.diff_pressure_read(&data_.diff_pressure, &data_.diff_pressure_temp, &data_.diff_pressure_velocity);
+    }
+    break;
+  case 2:
+    if (rf_.board_.sonar_present())
+    {
+      data_.sonar_range = rf_.board_.sonar_read();
+    }
+    break;
+  case 3:
+    if (rf_.board_.mag_present())
+    {
+      float mag[3];
+      rf_.board_.mag_read(mag);
+      data_.mag.x = mag[0];
+      data_.mag.y = mag[1];
+      data_.mag.z = mag[2];
+      correct_mag();
+    }
+    break;
   }
+  next_sensor_to_update_ = (next_sensor_to_update_ + 1) % 4;
```

Comment:
Yeah.  I do too.  It just seemed like a lot of overhead at the time.  Perhaps it would make it cleaner